pkg/reconciler/namespace/controller.go

```diff
@@ -0,0 +1,268 @@
+	// other workers.
+	defer c.queue.Done(key)
+
+	var err error
+	if strings.HasPrefix(key, "namespace::") {
+		err = c.processNamespace(ctx, strings.TrimPrefix(key, "namespace::"))
+	} else if strings.HasPrefix(key, "cluster::") {
+		err = c.processCluster(ctx, strings.TrimPrefix(key, "cluster::"))
+	} else {
+		err = c.processResource(ctx, key)
+	}
+	c.handleErr(err, key)
+	return true
+}
+
+func (c *Controller) handleErr(err error, key string) {
+	// Reconcile worked, nothing else to do for this workqueue item.
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	if _, ok := err.(*kerrors.RetryableError); ok {
+		num := c.queue.NumRequeues(key)
+		klog.Errorf("Error reconciling key %q, retrying... (#%d): %v", key, num, err)
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	// Give up and report error elsewhere.
+	c.queue.Forget(key)
+	runtime.HandleError(err)
+	klog.Errorf("Dropping key %q after non-retryable error: %v", key, err)
+}
+
+// key is gvr::namespace/name
+func (c *Controller) processResource(ctx context.Context, key string) error {
+	parts := strings.SplitN(key, "::", 2)
+	if len(parts) != 2 {
+		klog.Errorf("Error parsing key %q; dropping", key)
+		return nil
+	}
+	gvrstr := parts[0]
+	gvr, _ := schema.ParseResourceArg(gvrstr)
+	if gvr == nil {
+		klog.Errorf("Error parsing GVR %q; dropping", gvrstr)
+		return nil
+	}
+	key = parts[1]
+
+	obj, exists, err := c.ddsif.IndexerFor(*gvr).GetByKey(key)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		klog.Infof("object %q does not exist", key)
+		return nil
+	}
+
+	_, clusterAwareName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		klog.Errorf("splitting key %q: %v", key, err)
+		return err
+	}
+	clusterName, _ := clusters.SplitClusterAwareKey(clusterAwareName)
+	if clusterName != "admin" {
+		klog.Infof("Skipping update for non-admin cluster %q", clusterName)
+		return nil
+	}
+
+	current, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		klog.Errorf("object was not Unstructured: %T", obj)
+		return err
+	}
+
+	if namespaceBlocklist.Has(current.GetNamespace()) {
+		klog.V(2).Infof("Skipping syncing namespace %q", current.GetNamespace())
+		return nil
+	}
+
+	return c.reconcileResource(ctx, current, gvr)
+}
+
+// namespaceBlocklist holds a set of namespaces that should never be synced from kcp to physical clusters.
+var namespaceBlocklist = sets.NewString("kube-system", "kube-public", "kube-node-lease")
+
+func (c *Controller) processNamespace(ctx context.Context, key string) error {
+	_, clusterAwareName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		klog.Errorf("splitting key %q: %v", key, err)
+		return err
+	}
+	clusterName, name := clusters.SplitClusterAwareKey(clusterAwareName)
+	if clusterName != "admin" {
+		klog.Infof("Skipping update for non-admin cluster %q", clusterName)
+		return nil
+	}
+	if namespaceBlocklist.Has(name) {
+		klog.V(2).Infof("Skipping syncing namespace %q", name)
+		return nil
+	}
```

Comment:
Good idea. I don't know if it practically affects performance much, but we might as well filter as early as we can, rather than knowingly putting no-ops in the queue.